### PR TITLE
Add functions to train new word2vec model and loggers

### PR DESCRIPTION
**Pull request by florian-huber**
Add code to simply building a word2vec model suitable for spec2vec.
- ``train_new_word2vec_model`` function with set default parameters and more intuitive handling of learning rate decay.
- Callback function for logging (not pretty yet, I tried adding a ``tqdm`` progress bar but that gave me a whole bunch of Gensim errors).
- Callback function for saving the model at specified epochs.
- Tests.

### CODE REVIEW

File: spec2vec/model_building.py [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+from spec2vec.utils import TrainingProgressLogger
+
+
+def train_new_word2vec_model(documents: List, iterations: Union[List, int], filename: str = None, **kwargs):
+    """Train a new Word2Vec model (using gensim). Save to file if filename is given.
+
+    Parameters
+    ----------
+    documents:
+        List of documents, each document being a list of words (strings).
```
Comment (sverhoeven) [inline spec2vec/model_building.py:31]: I expected a list of documents where each document is a list of (word, count) tuples. 
In the case of spectrum's I expected the peak/loss and intensity tuple, but when take documents list from a test and do `list(documents[0])` I get 

```python
['peak@0.0',
 'peak@1.0',
 'peak@2.0',
 'peak@3.0',
 'peak@4.0',
 'peak@5.0',
 'peak@6.0',
 'peak@7.0',
 'peak@8.0',
 'peak@9.0']
```

How are counts or intensitiies being handled?

File: spec2vec/model_building.py [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+    # Set callbacks
+    callbacks = []
+    if settings["progress_logger"]:
+        training_progress_logger = TrainingProgressLogger(num_of_epochs)
+        callbacks.append(training_progress_logger)
+    if filename:
+        if isinstance(iterations, int):
+            iterations = [iterations]
+        model_saver = ModelSaver(num_of_epochs, iterations, filename)
+        callbacks.append(model_saver)
```
Comment (sverhoeven) [inline spec2vec/model_building.py:94]: Why not use [model.save(filename)](https://radimrehurek.com/gensim/models/word2vec.html#gensim.models.word2vec.Word2Vec.save)?

File: spec2vec/model_building.py [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+
+    # Train word2vec model
+    model = gensim.models.Word2Vec(documents, sg=settings["sg"], negative=settings["negative"],
+                                   size=settings["size"], window=settings["window"],
+                                   min_count=settings["min_count"], workers=settings["workers"],
+                                   iter=num_of_epochs, alpha=alpha, min_alpha=min_alpha,
+                                   seed=321, compute_loss=True, callbacks=callbacks)
```
Comment (sverhoeven) [inline spec2vec/model_building.py]: Hardcoding seed is a bit weird to me, as it already defaults to 1 which seem a good seed as well.

File: spec2vec/model_building.py [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+        training_progress_logger = TrainingProgressLogger(num_of_epochs)
+        callbacks.append(training_progress_logger)
+    if filename:
+        if isinstance(iterations, int):
+            iterations = [iterations]
+        model_saver = ModelSaver(num_of_epochs, iterations, filename)
+        callbacks.append(model_saver)
+
+    # Train word2vec model
+    model = gensim.models.Word2Vec(documents, sg=settings["sg"], negative=settings["negative"],
```
Comment (sverhoeven) [inline spec2vec/model_building.py]: Some **kwargs are stored in settings, but  gensim.models.Word2Vec has more arguments like `min_alpha`.

Could you add all **kwargs to `**settings` and use **settings as argument instead of all the `...=settings['...']?

Then any advanced Gensim parameters are settable by calling train_new_word2vec_model(). Also the docstring becomes easier by pointing to the parameters of  `py:class:gensim.models.word2vec.Word2Vec`.

File: spec2vec/model_building.py [hunk 5]
```diff
@@ -0,0 +1,129 @@
+from typing import List
+from typing import Tuple
+from typing import Union
+import gensim
+from spec2vec.utils import ModelSaver
+from spec2vec.utils import TrainingProgressLogger
+
+
+def train_new_word2vec_model(documents: List, iterations: Union[List, int], filename: str = None, **kwargs):
```
Comment (sverhoeven) [inline spec2vec/model_building.py]: Type return with
```suggestion
def train_new_word2vec_model(documents: List, iterations: Union[List, int], filename: str = None, **kwargs) -> gensim.models.Word2Vec:
```

File: spec2vec/model_building.py [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+    """Train a new Word2Vec model (using gensim). Save to file if filename is given.
+
```
Comment (sverhoeven) [inline spec2vec/model_building.py:15]: A example snippet how to use it would help me.

**Review by sverhoeven — CHANGES_REQUESTED** [review]

File: spec2vec/model_building.py (hunk 2 above)
Comment (florian-huber) [reply spec2vec/model_building.py:94]: ``model.save(filename)`` will work once the model is finished training.
I added the ``ModelSaver`` to allow the user (me....) to specify when to save the model during training. Say I add ``iterations=[5,10, 20, 30, 40, 50]`` then it will save a version of the model at all those specified iterations.

File: spec2vec/model_building.py (hunk 3 above)
Comment (florian-huber) [reply spec2vec/model_building.py]: True, the default was removed.

File: spec2vec/model_building.py (hunk 1 above)
Comment (florian-huber) [reply spec2vec/model_building.py:31]: Intensities are not used for training the word2vec model. They will only come into play when the document vectors are calculated. Multiple counts are rare except if a word occurs several times due to the rounding to 1,2 decimals. In that case the word will appear multiple times in the list.

File: spec2vec/model_building.py (hunk 4 above)
Comment (florian-huber) [reply spec2vec/model_building.py]: Good idea. I switched to **settings.
But I got a bit more complicated than I wanted because I on purpose use some arguments that deviate from gensim, namely ``learning_rate_initial`` and ``learning_rate_decay`` instead of ``alpha`` and ``min_alpha``. ``alpha``==``learning_rate_initial``, that could go, but ``min_alpha`` is for gensim the final learning rate which I would like to avoid because it has very non-intuitive consequences. That makes it a bit messy... I'll see if I can make it any cleaner.

File: spec2vec/model_building.py (hunk 1 above)
Comment (sverhoeven) [reply spec2vec/model_building.py:31]: OK

File: spec2vec/model_building.py (hunk 2 above)
Comment (sverhoeven) [reply spec2vec/model_building.py:94]: ah, I did not expect iterations to influence the save of the model to disk. 

Could you add something about iterations to the filename argument description?

File: spec2vec/model_building.py (hunk 6 above)
Comment (florian-huber) [reply spec2vec/model_building.py:15]: I added a snipped code example to the docstring.

File: spec2vec/model_building.py (hunk 4 above)
Comment (florian-huber) [reply spec2vec/model_building.py]: I again refactored it a bit. 
Not sure about the docstring though. I added ``py:class:gensim.models.word2vec.Word2Vec`` for the **settings argument. But it would still be good to keep all arguments in the docstring for which spec2vec has a certain default, right?

File: spec2vec/model_building.py (hunk 2 above)
Comment (florian-huber) [reply spec2vec/model_building.py:94]: I extended the description for filename accordingly.

Comment (sverhoeven) [reply spec2vec/model_building.py:94]: Great

File: spec2vec/model_building.py (hunk 6 above)
Comment (sverhoeven) [reply spec2vec/model_building.py:15]: nice thanks

File: spec2vec/model_building.py (hunk 4 above)
Comment (sverhoeven) [reply spec2vec/model_building.py]: Thanks, that a lot nicer.

Keep documenting the most used arguments if fine by me. When Gensim changes the arguments this docstring might become out of date, but I don't expect the oarguments listed in this docstring to change.

File: spec2vec/model_building.py [hunk 7]
```diff
@@ -0,0 +1,172 @@
...
+        iterations) using the pattern: filename_TEMP_{#iteration}epoch.model
+    learning_rate_initial:
+        Set initial learning rate. Default is 0.025.
+    learning_rate_decay:
+        After every epoch the learning rate will be lowered by the learning_rate_decay.
+        Default is 0.00025.
+    progress_logger:
+        If True, the training progress will be printed every epoch. Default is True.
+    **settings
+        See py:class:gensim.models.word2vec.Word2Vec.
```
Comment (sverhoeven) [inline spec2vec/model_building.py]:
```suggestion
        All other named arguments will be passed to the :py:class:`gensim.models.word2vec.Word2Vec` constructor.
```

**Review by sverhoeven — CHANGES_REQUESTED** [review]

**Comment by florian-huber** [comment]
@sverhoeven Thanks for the review so far. I just also extended the readme with a more extensive code example (wasn't part of this PR before... but it seemed to fit).

**Review by sverhoeven — APPROVED** [review]

**Comment by florian-huber** [comment]
Thanks a lot for the review!
